PHP SDK v2.0.0 (#1083)
rudderlabs/rudderstack-docs: docs/sources/event-streams/sdks/rudderstack-php-sdk.mdx

description: Use RudderStack’s PHP SDK to send your server-side events to various destinations.
---

RudderStack's PHP SDK lets you track and send the events from your PHP applications to the specified destinations.

<div class="infoBlock">
Refer to the PHP SDK's <a href="https://github.com/rudderlabs/rudder-php-sdk">GitHub codebase</a> for the implementation-specific details.
</div>

<GhBadge
  url={'https://packagist.org/packages/rudderstack/rudder-php-sdk'}
  repo={'packagist/v/rudderstack/rudder-php-sdk'}
  style={'flat'}
/>

<div class="infoBlock">
RudderStack's PHP SDK supports PHP version 7.4 and above.

> </div>

Reviewer: Can we club this info block with the one above? 

```
RudderStack's PHP SDK supports PHP version 7.4 and above. Refer to the SDK's <a href="https://github.com/rudderlabs/rudder-php-sdk">GitHub codebase</a> for the implementation-specific details.
```

## SDK setup requirements

1. Sign up to [RudderStack Cloud](https://app.rudderstack.com/signup).
2. <Link to="/dashboard-guides/sources/#adding-a-source">Set up a PHP source</Link> in your dashboard. You should be able to see the <strong>write key</strong> for this source, as shown:

<img src="../../../assets/event-stream-sources/php-write-key-new.png" alt="PHP source write key" />

> You will also need the data plane URL. Refer to the <Link to="/resources/glossary/#data-plane-url">Glossary</Link> for more information on the data plane URL and where to find it.

Reviewer: ```suggestion
You will also need the data plane URL. Refer to the <Link to="/dashboard-guides/overview/#data-plane-url">Dashboard Overview</Link> guide for more information on the data plane URL and where to find it.
```

Reviewer: Best to cross-reference the dashboard guide instead of glossary, and keeping it consistent wrt. standardization.

To initialize the RudderStack client, run the following code snippet:

```php
use Rudder\Rudder;

require_once realpath(__DIR__ . '/vendor/autoload.php');

Rudder::init(WRITE_KEY, array(
  "data_plane_url" => DATA_PLANE_URL,
  "consumer"       => "lib_curl",
  "debug"          => false,
  "max_queue_size" => 10000,
  "flush_at"       => 100,
  "ssl"            => true
));
```

<div class="infoBlock">

> RudderStack accepts the <code class="inline-code">DATA_PLANE_URL</code> with or without the protocol. If the protocol is missing, RudderStack automatically prepends it to the URL based on whether the <code class="inline-code">ssl</code> option is set to <code class="inline-code">true</code> or  <code class="inline-code">false</code>.

Reviewer: ```suggestion
RudderStack accepts the <code class="inline-code">DATA_PLANE_URL</code> with or without the protocol. If the protocol is missing, RudderStack automatically prepends it to the URL based on whether the <code class="inline-code">ssl</code> option is set to <code class="inline-code">true</code> or <code class="inline-code">false</code>.
```

| `false` | `example.dataplane.com`         |

<div class="warningBlock">
While initializing the RudderStack client, the <code class="inline-code">ssl</code> field is optional if you are using the HTTPS protocol. If included, its value must be set to <code class="inline-code">true</code>. In case of the HTTP protocol, <code class="inline-code">ssl</code> is a required field and must always be set to <code class="inline-code">false</code>.
</div>

> ### Other initialization options

Reviewer: I think we are listing all the initialization options in this section. Should we name the section as **SDK initialization options** instead?

> RudderStack provides some additional initialization options which are explained in the following table:

Reviewer: ```suggestion
The RudderStack PHP SDK provides the following initialization options:
```

> | Option        | Data type         | Description     |

Reviewer: Should we specify which options are required and optional?

Author: All are optional :)

| :----------- | :--------------- | :----------|
|  `consumer`           | String      | Explicitly marks which consumer to use. The default value is `lib_curl`. Refer to the <Link to="#consumers">Consumers</Link> section for more information. |

> |  `data_plane_url `       | String | Explicitly sets the URL to data plane. The default value is `hosted.rudderlabs.com`.  |

Reviewer: ```suggestion
|  `data_plane_url `       | String | Explicitly sets the data plane URL. The default value is `hosted.rudderlabs.com`.  |
```

|  `debug`          | Boolean  |  Determines whether to log messages and wait for a response. The default value is `false`. It makes the queuing consumers non-async and blocks the library until a response is received from the API. Hence it is **not recommended** for production use.  |
|  `ssl`           | Boolean      | Determines whether to use TLS instead of SSL for the socket connection. The default value is `false`. |

> |  `error_handler`       | Function | Function to handle errors and is particularly useful for debugging. Note that if the debug mode is not specified, then the `error_handler` is only called on connection level errors and timeouts. For example, `function ($code, $message) {}`   |

Reviewer: ```suggestion
|  `error_handler`       | Function | Function to handle errors and is particularly useful for debugging. Note that if the debug mode is not specified, then the `error_handler` is only called on connection level errors and timeouts. For example, `function ($code, $message) {}`.   |
```

> |  `max_queue_size`           | Number      | Max size of the queue. The default value is 10000 items. |

Reviewer: ```suggestion
|  `max_queue_size`           | Number      | Max size of the queue. The default value is 10000. |
```

> |  `batch_size`       | Number | Items to be sent in a single curl request. The default value is 100. It is recommended to use `flush_at` instead of `batch_size` as this option is deprecated.   |

Reviewer: Should we use `Integer` as the data type instead of `Number`, as that is the standard convention?

|  `flush_at`           | Number      | Items to be sent in a single curl request. The default value is 100. |
|  `timeout `       | Number | Number of seconds to wait for the socket request to time out. The default value is 0.5.  |

> |  `filename`           | String      |Location to write the log file. The default value is `/tmp/analytics.log` when file consumer is selected. |

Reviewer: ```suggestion
|  `filename`           | String      |Location to write the log file. The default value is `/tmp/analytics.log` when the file consumer is selected. |
```

> |  `compress_request`       | Boolean | Determines whether to use gzipped request payloads. The default value is `true`. It is supported for server versions 1.4.0 and above.  |

Reviewer: ```suggestion
|  `compress_request`       | Boolean | Determines whether to use gzipped request payloads. The default value is `true`. It is supported for <a href="https://github.com/rudderlabs/rudder-server">RudderStack server</a> v1.4.0 and above.  |
```

|  `flush_interval`           | Number      | Frequency in milliseconds to send data using flush method execution. The default value is 10000 miliseconds. |

> |  `curl_timeout `       | Number | Timeout for the curl connections. The default value is 0, meaning infinite. |

Reviewer: ```suggestion
|  `curl_timeout `       | Number | Timeout for the curl connections. The default value is 0, that is, infinite. |
```

|  `curl_connecttimeout`           | Number      | Connect timeout for the curl connections. The default value is 300. |

> |  `max_item_size_bytes`       | Number | Maximum message item size. The default value is 32000 bytes (32kb). |

Reviewer: ```suggestion
|  `max_item_size_bytes`       | Number | Maximum message item size. The default value is 32000 Bytes (32KB). |
```

> |  `max_queue_size_bytes`           | Number      | Maximum batch size. The default value is 512000 bytes (512kb).   |

Reviewer: ```suggestion
|  `max_queue_size_bytes`           | Number      | Maximum batch size. The default value is 512000 Bytes (512KB).   |
```

> |  `filepermissions `       | String | File permissions for file consumer. The default value is 0644.   |

Reviewer: Do we need to add any specific reference for more information? Wondering where we got `0644` from.

Author: @bardisg Is there any more info we can add here?

## Migrating from v1 to v2

> To migrate to PHP SDK v2, ensure the following conditions are met:

Reviewer: ```suggestion
To migrate to PHP SDK v2, make sure the following conditions are met:
```

- Your PHP version should be v7.4 or above.

> - If you are using self hosted <a href="https://github.com/rudderlabs/rudder-server">rudder-server</a>, either upgrade it to v1.4.0/above or turn off the gzipped request payloads  (by setting `compress_request` to `false`).

Reviewer: ```suggestion
- If you are using self-hosted <a href="https://github.com/rudderlabs/rudder-server">rudder-server</a>, either upgrade it to v1.4.0 or above, or disable the gzipped request payloads  (by setting `compress_request` to `false`).
```

## Sending events from the RudderStack client

<div class="warningBlock">
<strong>RudderStack does not store or persist the user state in any of the server-side SDKs</strong>. <br /><br />Unlike the client-side SDKs that deal with a single user at a given time, the server-side SDKs deal with multiple users simultaneously. Therefore, you must specify either the <code class="inline-code">userId</code> or <code class="inline-code">anonymousId</code> <strong>every time</strong> while making any API calls supported by the PHP SDK.
</div>

## Identify

The <Link to="/event-spec/standard-events/identify/">`identify`</Link> call lets you identify a visiting user and associate them to their actions. It also lets you record the traits about them like their name, email address, etc.

A sample `identify` call made using the PHP SDK is shown below:

```php
Rudder::identify(array(
  "userId" => "1hKOmRA4GRlm",
  "traits" => array(
    "email" => "alex@example.com",
    "name" => "Alex Keener",
    "friends" => 25
  )
));
```

The `identify` method parameters are as described below:

| **Field**      | **Type** | **Description**        |
| :------------- | :------- | :------------- |

> | `anonymousId` <br/> <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>userId</strong> is absent.</span> | String   | Sets the user ID for cases where there is no unique identifier for the user. |

Reviewer: ```suggestion
| `anonymousId` <br/> <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>userId</strong> is absent.</span> | String   | The SDK automatically sets this identifier in cases where there is no unique identifier for the user. |
```

Reviewer: Standardizing the descriptions as done in the Java SDK doc.

| `userId` <br/> <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>anonymousId</strong> is absent.</span>      | String   |  Unique identifier for a particular user in your database.                                                                  |

> | `context`      | Object   |  Dictionary of information that provides context about a message. However, it is not directly related to the API call.      |

Reviewer: ```suggestion
| `context`      | Object   |  An optional dictionary of information that provides context about a message. It is not directly related to the API call.      |
```

> | `integrations` | Object   |  A dictionary containing the destinations to be either enabled or disabled.     |

Reviewer: ```suggestion
| `integrations` | Object   | An optional dictionary containing the destinations to be either enabled or disabled.     |
```

| `timestamp`    | Date     | The timestamp of the message's arrival.        |

> | `traits`       | Object   | Dictionary of the traits associated with the user, such as `name` or `email`.    |

Reviewer: ```suggestion
| `traits`       | Object   | Dictionary of the user's traits like `name` or `email`.    |
```

The `track` method parameters are as described below:

| Name           | Type   | Description     |
| :------------- | :----- | :------------------------------- |

> | `user_id` <br/> <span style="color: #4D4DFF;font-size:12px;">Required</span>     | String |  The developer identification for your user.                                                                                 |

Reviewer: ```suggestion
| `userId` <br/> <span style="color: #4D4DFF;font-size:12px;">Required</span>     | String |  Unique identifier for a user in your database.    |
```

Reviewer: Need confirmation - the snippet says `userId` and the table has `user_id`. @bardisg

Reviewer 2: Did not check this section as I did not affect anything there. This issue was there. let me check all key names accross the doc even though I did not change them.

Correct key names are:
- userId
- anonymousId

They are wrong only on the track event section

> | `event` <br/> <span style="color: #4D4DFF;font-size:12px;">Required</span>        | String |  Name of the event being performed by the user.                                                                              |

Reviewer: ```suggestion
| `event` <br/> <span style="color: #4D4DFF;font-size:12px;">Required</span>        | String |  Name of the event.                                                   |
```

> | `event` <br/> <span style="color: #4D4DFF;font-size:12px;">Required</span>        | String |  Name of the event being performed by the user.                                                                              |

Reviewer: ```suggestion
| `anonymousId` <br/> <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>userId</strong> is absent.</span> | String   | The SDK automatically sets this identifier in cases where there is no unique identifier for the user. |
| `event` <br/> <span style="color: #4D4DFF;font-size:12px;">Required</span>        | String |  Name of the event being performed by the user.          |
```

Reviewer: Need confirmation - should it be `anonymousId` or `anonymous_id`?

Author: @bardisg can you please confirm

Reviewer 2: anonymousId

> | `properties`   | Object |  Dictionary of the properties associated with a particular event.                                                           |

Reviewer: ```suggestion
| `properties`   | Object |  An optional dictionary of the properties associated with the event.          |
```

> | `context`      | Object |  Dictionary of information that provides context about a message. However, it is not directly related to the API call.      |

Reviewer: ```suggestion
| `context`      | Object |  An optional dictionary of information that provides context about the message. It is not directly related to the API call.      |
```

| `timestamp`    | Date   |  The timestamp of the message's arrival.                                                                                    |

> | `anonymous_id` | String |  Sets the user ID for cases where there is no unique identifier for the user. Either `userId` or `anonymousId` is required. |

Reviewer: ```suggestion
```

Reviewer: Would be better to have user ID followed by anonymous ID.

> | `integrations` | Object |  A dictionary containing the destinations to be either enabled or disabled.                                                 |

Reviewer: ```suggestion
| `integrations` | Object |  An optional dictionary containing the destinations to be enabled or disabled.                |
```

  "category" => "Docs",
  "name" => "PHP library",
  "properties" => array(
    "url" => "https://rudderstack.com/libraries/php/"
  )
));
```

The `page` method parameters are as described below:

| **Field**      | **Type** | **Description**                  |
| :------------- | :------- | :-------------------------------- |

> | `anonymousId` <br/> <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>userId</strong> is absent.</span> | String   |  Sets the user ID for cases where there is no unique identifier for the user.  |

Reviewer: ```suggestion
| `anonymousId` <br/> <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>userId</strong> is absent.</span> | String   |  The SDK automatically sets this identifier in cases where there is no unique identifier for the user.  |
```

| `userId`   <br/> <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>anonymousId</strong> is absent.</span>    | String   | Unique identifier for a particular user in your database.      |

> | `context`      | Object   |  Dictionary of information that provides context about a message. However, it is not directly related to the API call.      |

Reviewer: ```suggestion
| `context`      | Object   |  An optional dictionary of information that provides context about the event. It is not directly related to the API call.      |
```

> | `integrations` | Object   |  A dictionary containing the destinations to be either enabled or disabled.                                                 |

Reviewer: ```suggestion
| `integrations` | Object   |  An optional dictionary containing the destinations to be enabled or disabled.        |
```

> | `name`  <br/> <span style="color: #4D4DFF;font-size:12px;">Required</span>        | String         | Name of the page being viewed.                                                                                             |

Reviewer: ```suggestion
| `name`  <br/> <span style="color: #4D4DFF;font-size:12px;">Required</span>        | String         | Name of the viewed page.      |
```

> | `properties`   | Object   |  Dictionary of the properties associated with the page being viewed, such as `url` and `referrer`.                          |

Reviewer: ```suggestion
| `properties`   | Object   |  An optional dictionary of the properties associated with the viewed page, like `url` and `referrer`.       |
```

The `screen` method parameters are as described below:

| **Field**      | **Type** | **Description**                 |
| :------------- | :------- |  :---------------------------- |

> | `anonymousId` <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>userId</strong> is absent.</span> | String   |Sets the user ID for cases where there is no unique identifier for the user. Either `userId` or `anonymousId` is required. |

Reviewer: ```suggestion
| `anonymousId` <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>userId</strong> is absent.</span> | String   | The SDK automatically sets this identifier in cases where there is no unique identifier for the user. |
```

> | `userId`  <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>anonymousId</strong> is absent.</span>     | String   | Unique identifier for a particular user in your database.            |

Reviewer: ```suggestion
| `userId`  <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>anonymousId</strong> is absent.</span>     | String   | Unique identifier for a user in your database. |
```

> | `context`      | Object   |  Dictionary of information that provides context about a message. However, it is not directly related to the API call.      |

Reviewer: ```suggestion
| `context`      | Object   |  An optional dictionary of information that provides context about the event. It is not directly related to the API call.  |
```

> | `integrations` | Object   | A dictionary containing the destinations to be either enabled or disabled.                                                 |

Reviewer: ```suggestion
| `integrations` | Object   | An optional dictionary containing the destinations to be enabled or disabled.           |
```

> | `name` <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>userId</strong> is absent.</span>        | String                                  | Name of the screen being viewed.                                                                                           |

Reviewer: ```suggestion
| `name` <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>userId</strong> is absent.</span>        | String       | Name of the viewed screen.   |
```

> | `properties`   | Object   | Dictionary of the properties associated with the page being viewed, such as `url` and `referrer`                           |

Reviewer: ```suggestion
| `properties`   | Object   | An optional dictionary of the properties associated with the screen, like `url` and `referrer`.                           |
```

The `alias` method parameters are as mentioned below:

| **Field**      | **Type** | **Description**          |
| :------------- | :------- | :---------------------------------------- | :--------------------------- |
| `userId`  <br/> <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>anonymousId</strong> is absent.</span>     | String   | Unique identifier for a particular user in your database.                                                             |
| `context`      | Object   | Dictionary of information that provides context about a message. However, it is not directly related to the API call. |
| `integrations` | Object   | A dictionary containing the destinations to be either enabled or disabled.                                            |

> | `previousId` <br/> <span style="color: #4D4DFF;font-size:12px;">Required</span>  | String    | The previous unique identifier of the user.                                                                           |

Reviewer: We should move the `previousId` parameter after `userId` as these are the two required parameters.

| `traits`       | Object   | Dictionary of the properties or traits associated with the group, such as `email` or `name`.                          |
| `timestamp`    | Date     | The timestamp of the message's arrival.                                                                               |

<div class="infoBlock">
For a detailed explanation of the <code class="inline-code">alias</code> call, refer to our <a href="https://rudderstack.com/docs/event-spec/standard-events/alias">RudderStack API Specification</a> guide.
</div>

## Consumers

> You can specify different consumers to make requests to RudderStack servers as explained below:

Reviewer: ```suggestion
You can specify the following consumers to make requests to RudderStack:
```

Reviewer: IMO we can just say `RudderStack` instead of `RudderStack servers` for simplicity. Thoughts? Can make the change in the following sections if we are on the same page.

### Lib-Curl Consumer

The [lib-curl consumer](https://github.com/rudderlabs/rudder-php-sdk/blob/develop/lib/Consumer/LibCurl.php) is RudderStack's default PHP library. You can use it for faster response during lighter loads. It runs synchronously, queuing and sending calls in batches to RudderStack's servers. 

### Fork-Curl Consumer

> The [fork-curl consumer](https://github.com/rudderlabs/rudder-php-sdk/blob/develop/lib/Consumer/ForkCurl.php) can be used when you can't use persistent sockets or want to ensure quick response for lighter loads. It creates an in-memory queue which buffers the `identify` and `track` calls. The queue is flushed by forking an async `curl` process that sends a batch request. It happens after every `100` calls or at the end of serving the page, by default. 

Author: This [segment doc](https://segment.com/docs/connections/sources/catalog/libraries/server/php/#fork-curl-consumer) is the only source for this info. 

Maybe @bardisg can confirm more. Should we add more info about the configurability of this consumer's value, as Amey suggested above?

> The [fork-curl consumer](https://github.com/rudderlabs/rudder-php-sdk/blob/develop/lib/Consumer/ForkCurl.php) can be used when you can't use persistent sockets or want to ensure quick response for lighter loads. It creates an in-memory queue which buffers the `identify` and `track` calls. The queue is flushed by forking an async `curl` process that sends a batch request. It happens after every `100` calls or at the end of serving the page, by default. 

Reviewer 2: There is a documented property flush_at already to configure this

### Socket Consumer

> You can use the [socket consumer](https://github.com/rudderlabs/rudder-php-sdk/blob/develop/lib/Consumer/Socket.php) which lets you make requests to RudderStack when you can't spawn other processes from your PHP scripts. It initiates  a socket request to RudderStack servers each time an `identify` or `track` call is made. The socket request writes the event data and closes the connection before waiting for a response. If your servers are dealing with more than 100's of requests per second or cannot use a persistent connection, it is recommended use some other consumer.

Reviewer: ```suggestion
You can use the [socket consumer](https://github.com/rudderlabs/rudder-php-sdk/blob/develop/lib/Consumer/Socket.php) which lets you make requests to RudderStack when you can't spawn other processes from your PHP scripts. It initiates  a socket request to RudderStack servers each time an `identify` or `track` call is made. The socket request writes the event data and closes the connection before waiting for a response. If your servers are dealing with hundreds of requests per second or cannot use a persistent connection, it is recommended use some other consumer.
```

### File Consumer

You can also use the [file consumer](https://github.com/rudderlabs/rudder-php-sdk/blob/develop/lib/Consumer/File.php) for making requests to RudderStack. It records each `identify` or `track` call you make to a log file. Further, it uploads the log file by running the `file.php` file in RudderStack's [github repository](https://github.com/rudderlabs/rudder-php-sdk/blob/develop/lib/Consumer/File.php).

You can run the `examples/SendBatchFromFile.php` file to upload your log file to RudderStack, as shown:

```bash
php examples/SendBatchFromFile.php --secret YOUR_WRITE_KEY --file /tmp/analytics.log
```

<div class="infoBlock">

> It is recommended to run this command as a cron job after every few minutes so that your log files are created in manageable sizes. Every time you run the command, it removes the old log files once they are processed successfully.

Reviewer: ```suggestion
It is recommended to run this command as a CRON job after every few minutes so that your log files are created in manageable sizes. Every time you run the command, it removes the old log files once they are processed successfully.
```

</div>

> You can create a new cron job to upload your log files. A sample snippet to create a cron job is as shown:

Reviewer: ```suggestion
You can create a new CRON job to upload your log files, as shown:
```

```bash
$ # create a cron job that runs as www-data every minute

> $ echo '*/1 * * * * www-data php /my/path/to/analytics-php/send.php > /dev/null' | sudo tee /etc/cron.d/analytics

Reviewer 2: SendBatchFromFile.php insead of send.php
